### DISCUSSION TITLE
Trying to use tinyUSB on an STM32F7

### DISCUSSION THREAD

**Discussion by Chicowolf** [hathach/tinyusb#1591] [Q&A] [answered]
Hello everybody.

I'm trying to use tinyUSB on a Nucleo-F767ZI, following this tutorial: https://anvilelectronics.ovh/tinyusb-na-stm32-device-cdc/
After following all the steps of this guy, when I try to build the source code in CubeIDE I get more than 20 error messages. All the messages complain about multiple definitions, for example:

**_multiple definition of `dcd_edpt_clear_stall'; ./Core/tinyUSB/portable/st/synopsys/dcd_synopsys.o:D:/Progetti/STM32CubeIDE/testUSB/Debug/../Core/tinyUSB/portable/st/synopsys/dcd_synopsys.c:875: first defined here_**

I'd like to build the example for the CDC protocol and then move to the AUC2 example, but I'm stuck.
Can anyone help me configure this library correctly on CubeIDE?

**Comment by pigrew** [comment, marked as answer]
It sounds like you are perhaps including the TinyUSB project twice, or enabled multiple drivers.

Include more of the error log... where was the first definition of `dcd_edpt_clear_stall`?

**Reply by Chicowolf** [reply]
CubeIDE's error output is not quite clear, but it seems that the first declaration is in `portable/st/synopsys/dcd_synopsys.c`, while the error is in `portable/synopsys/dwc2/dcd_dwc2.c`

**Reply by pigrew** [reply]
That MCU is supported by two different drivers. My suggestion is to disable the compilation of one of those two source files in your project.

Looking at the source, I don't see a `#define` that can be used to disable one at a time. In the existing project, the makefiles are including one source or the other.

**Reply by Chicowolf** [reply]
I tried to remove the folder `dwc2` from `portable/synopsys` and it manages to compile, but the execution get stuck at line 504 inside `dcd_synopsys.c`:

>   while ((usb_otg->GRSTCTL & USB_OTG_GRSTCTL_AHBIDL) == 0U) {}

in particular `usb_otg->GRSTCT` is equal to zero.

On the other hand, if I remove the folder `synopsys` inside `portable/st`, it also compiles, but the execution get stuck at line 467 inside `dcd_dwc2.c`:

>   TU_ASSERT(gsnpsid == DWC2_OTG_ID || gsnpsid == DWC2_FS_IOT_ID || gsnpsid == DWC2_HS_IOT_ID);

in particular `gsnpsid` is equal to zero.

Why does it get stuck?

**Reply by pigrew** [reply]
The dwc2 driver is newer, and likely to be preferred in the future. It looks like the STM32f7 doesn't have the `gsnpsid` register.  I'm not sure if the driver has been tested on the F7 (though it did get a checkmark on the supported MCU page..). (I've never used this driver....)

Perhaps comment out that check in the dwc2 driver, like was done with the GD32VF103. If that patch works, someone should add a compiler macro to remove all references to that register for the indicated platforms.

**Reply by Chicowolf** [reply]
I did get what wasn't working. The problem was that I downloaded the CDC example files from the master branch, but apparently they don't work with the release version of the library.
Now the library works with my Nucleo-F767ZI and Nucleo-H745ZI.
Thank you for your time and patience :)

**Reply by hathach** [reply]
thank you Nathan for helping, for stm32 dwc2 controller, if gsnpsid == 0, it is indicattion that usb controller clock is not enabled.
